### ISSUE TITLE
Promise.andThen signature with disjunction types

### ISSUE THREAD

**Issue by bblfish** [scala-js/scala-js-dom#164]
Currently `andThen` is defined

```
  def andThen[B](onFulfilled: js.Function1[A,B]): Promise[Any] = js.native
```

This loss of type in the response (it's of type `Promise[Any]` rather than `Promise[B]` makes it awkward  when writing functions using the `Fetch` api where one would like to write things like 

```
 def fetchListener(e: FetchEvent) = e.respondWith {
       fetch(e.request.url).andThen{
             response=> f(response),
             err => err 
      }
  }
``` 
where respondWith takes a `Promise[Response]`. With the current signature one needs to add an extra cast.

Instead the following signature seems more correct and closer to the JS documentation [I have read](http://www.html5rocks.com/en/tutorials/es6/promises/)

```
def andThen[B](onFulfilled: js.Function1[A,B]|js.Function1[A,Promise[B]]): Promise[B] = js.native
```

Here we use the disjunction type since `andThen` can either take a function to a value or a function to something Promisy.

**Comment by mseddon** [comment]
Hmm. I've thought a little more about it and wouldn't a more correct typing would be:
```scala
def andThen[B](onFulfilled: js.Function1[A,B|Promise[C]]): Promise[B|C] = js.native
```

Since you could have a promise that conditionally returns either a value or a promise?

**Comment by bblfish** [comment]
yes, you reveal an interesting possibility opened by the andThen function, namely that it can pass a function which on some input would return a function that returns immediately or a function that returns a promise that may be running on some other worker thread perhaps...

Note you'd either 
1. need to put the `C` in the `andThen` method so that it is written andThen[B,C]...` 
2. you could have one type variable like this: def andThen[B](onFulfilled: js.Function1[A,B|Promise[B]]): Promise[B] = js.native
   this would then mean that users of this function that actually had a function that returns a `Promise[C]` or a `D` could use the the function like this 
```
fetch(url).andThen[B|D]{ 
  case Response(401,headers) if headers.contains(`WWW-Authenticate`) => fetch(signed(url))
  case r@Response(200,_) => r 
}
```

**Comment by chandu0101** [comment]
reason for current API returns `[Any]`  : https://github.com/scala-js/scala-js-dom/pull/101#discussion_r26439753

**Comment by bblfish** [comment]
Ah ok, I propose to add that information about potentially infinitely nested Promises to the method documentation, as that is something that most developers would rarely end up working out for themselves. ( at least not run of the mill scala devs... )

**Comment by mseddon** [comment]
Promises will likely be provided by the Scala.js core library in the next release (and include slightly better typing than currently provided, with caveats).  You can track the progress [here](https://github.com/scala-js/scala-js/pull/2209).

**Comment by danielwegener** [comment]
@mseddon  [Now they are](https://www.scala-js.org/news/2016/02/12/announcing-scalajs-0.6.7/). Should we directly migrate scala-js-dom to scalajs-0.6.7 or should we keep the scala-js-dom Promise until a later "major" release?

**Comment by sjrd** [comment]
I think we should immediately migrate to `js.Promise`, then immediately publish 0.9.0.

**Comment by mseddon** [comment]
@danielwegener I strongly agree with @sjrd here, we want to encourage people to move onto the core library as soon as possible without confusion.

**Comment by danielwegener** [comment]
Cool, would be willing to do that, need a short break from slide preparations :)

**Comment by mseddon** [comment]
:) nice one.  Thanks!

**Comment by danielwegener** [comment]
There you go: #211

**Comment by sjrd** [comment]
Fixed in #211.
